### PR TITLE
Sockets - Carla

### PR DESCRIPTION
**Pull request by carlabosco**
# Hotel
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
What was a design challenge that you encountered on this project? | Getting started and figuring out the classes and how they would interact was challenging. Even though I spent some time planning on paper and sketching I feel like I still need a lot more practice.
What was a design decision you made that changed over time over the project? | At first I wasn't convinced that an "administrative" class overlooking rooms and reservations would be necessary, but I soon felt the need to implement it.
What was a concept you gained clarity on, or a learning that you'd like to share? | Testing is starting to feel more natural, although I'm disappointed with the fact that I didn't always implemented tests first. 
What is an example of a _nominal test_ that you wrote for this assignment? What makes it a nominal case? | The tests I used for checking if my class objects can be instantiated are examples of nominal cases because they check for the basic use of the initialize method.
What is an example of an _edge case test_ that you wrote for this assignment? What makes it an edge case? | I tested for booking 21 rooms for the same date range. Booking 100 rooms would be a clear fail, and booking up to 20 would be a clear pass, so I believe that booking 21 is an example of edge case.
How do you feel you did in writing pseudocode first, then writing the tests and then the code? | I did this in the beginning of the project and it was great. Pseudocoding methods for the outcomes I would like to see gave me some clarity on the project as a whole.

### CODE REVIEW

File: lib/date_range.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
+require 'time'
+
+class DateRange
+  # Check if given date is valid 
+  def self.valid_date?(date)
+    begin
+      Date.parse(date, '%Y-%m-%d')
+      true
+    rescue => e
+      false
```
Comment (droberts-sea) [inline lib/date_range.rb:10]: I would prefer an explicit `return` here.

File: lib/date_range.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    end
+  end
+
+  def self.generate_date_range(check_in, check_out)
+    (check_in..check_out).select { |d|  valid_date?(d) }
+  end
+
+  def self.dates_overlap?(range1, range2)
+    range1.last > range2.first && range1.first < range2.last
+  end
```
Comment (droberts-sea) [inline lib/date_range.rb:20]: Good work getting this tricky logic figured out.

File: lib/date_range.rb [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+      Date.parse(date, '%Y-%m-%d')
+      true
+    rescue => e
+      false
+    end
+  end
+
+  def self.generate_date_range(check_in, check_out)
+    (check_in..check_out).select { |d|  valid_date?(d) }
+  end
```
Comment (droberts-sea) [inline lib/date_range.rb:16]: I'm not sure what the call to `valid_date?` is accomplishing. You're pretty much guaranteed that the dates generated by the range (`check_in..check_out`) are valid, since that's part of the Ruby language.

A more useful check might be that the checkout date comes after the checkin date (the range will not produce an error in this case).

File: lib/booking_central.rb [hunk 4]
```diff
@@ -0,0 +1,92 @@
+#require_relative 'room'
+require_relative 'date_range'
+require_relative 'reservation'
+require_relative 'block'
+
+class BookingCentral 
+  attr_accessor :rooms, :all_reservations, :blocks
+
```
Comment (droberts-sea) [inline lib/booking_central.rb:8]: You should be wrapping these classes in a module.

File: lib/booking_central.rb [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+  def assign_room(check_in, check_out)
+    random_available_room = list_available_rooms(check_in, check_out).sample
+    return random_available_room
+  end
+
+  def reserve_room(check_in:, check_out:, room: assign_room(check_in, check_out))
+    if list_available_rooms(check_in, check_out) == []
+      raise ArgumentError, "There is no availability for the dates provided."
+    else
```
Comment (droberts-sea) [inline lib/booking_central.rb]: This method ends up calling `list_available_rooms` up to 3 times, once here and once or twice via `assign_room`. A slightly cleaner implementation might be:

```ruby
def reserve_room(check_in:, check_out:, room: nil)
  room ||= assign_room(check_in, check_out)
  unless room
    raise ArgumentError, "There is no availability for the dates provided."
  end
  new_reservation = Reservation.new(check_in: check_in, check_out: check_out, room: room)
  # ... etc ...
end
```

File: lib/booking_central.rb [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+    end
+  end
+
+  def block_rooms(check_in:, check_out:, number_of_rooms:, rooms:, discount_rate:)
+    available_rooms = list_available_rooms(check_in, check_out)
+    number_of_rooms.times do |i|
+      room = assign_room(check_in, check_out)
+      @rooms << room
+      @rooms.map{ |room| room.blocked == true }
+    end
```
Comment (droberts-sea) [inline lib/booking_central.rb:55]: Line 58 isn't doing anything. You call `.map`, which is supposed to turn one array into another, but you don't save the result in a variable. Inside the block you use `==`, which checks for equality rather than doing assignment.

When choosing an enumerable, make sure you understand what problem you're trying to solve, and how the pattern of that enumerable solves it.

Comment (droberts-sea) [reply lib/booking_central.rb:55]: Also, I'm not sure that interacting with `@rooms` here makes sense - you should probably be creating a new empty array as a local variable and keeping track of rooms in that instead.

File: lib/booking_central.rb [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+    matching_reservations = @all_reservations.select { 
+      |reservation, details| reservation if DateRange.dates_overlap?(reservation.date_range, asking_date)
+    }
+    return matching_reservations
+  end
+
+  def list_available_rooms(check_in, check_out)
+    asking_date = DateRange.generate_date_range(check_in, check_out)
+    available_rooms = @rooms - (reservations_by_date(check_in, check_out)).map{ |reservation| reservation.room }
+    available_rooms -= @rooms.select{ |room| room.blocked == true }
```
Comment (droberts-sea) [inline lib/booking_central.rb]: First, I love that you've broken this logic out as a separate method, that makes the rest of your code much easier to read.

Second, good work getting this tricky logic with the lists of rooms sorted out.

Third, this logic is complex enough (particularly line 30) that it would probably make sense to split it across multiple lines.

File: lib/room.rb [hunk 8]
```diff
@@ -0,0 +1,10 @@
+class Room 
+  attr_accessor :number, :rate, :room_reservations, :blocked
+
+  def initialize(number, room_reservations:, blocked: false)
+    @number = number
+    @room_reservations = []
+    @blocked = blocked
```
Comment (droberts-sea) [inline lib/room.rb]: You have an instance variable here to keep track of the reservations for this room, but as far as I can tell you never use it.

Comment (droberts-sea) [reply lib/room.rb]: I'm not sure it makes sense to store `blocked` as a true or false value for the room. Whether or not a room is blocked depends on what date you're asking about, same as whether or not it's reserved. If you're keeping track of that here, the way to do it would be by storing a list of blocks that contain this room, and having a method `blocked(date_range)` that returns true or false.

File: spec/booking_central_spec.rb [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+      new_booking = booker.reserve_room(check_in: '2019-01-03', check_out: '2019-01-04', room: 1)
+
+      expect{(new_booking = booker.reserve_room(check_in: '2019-01-03', check_out: '2019-01-04', 
+        room: assign_room(check_in, check_out)).room)}.wont_equal 1
+    end
+
+    it 'rejects hotel overbooking' do
+      booker = BookingCentral.new
+      expect{21.times{booker.reserve_room(check_in: '2019-01-03', check_out: '2019-01-04', 
+        room: booker.assign_room('2019-01-03', '2019-01-04'))}}.must_raise ArgumentError
```
Comment (droberts-sea) [inline spec/booking_central_spec.rb:55]: Good work on this test case!

File: spec/date_range_spec.rb [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+  end
+
+  it 'rejects invalid date' do
+    expect(DateRange.valid_date?('2019-13-01')).must_equal false
+  end
+
+  it 'checks for overlaping date ranges' do
+    expect(DateRange.dates_overlap?(['2019-05-01', '2019-05-02', '2019-05-03'], ['2019-05-01', '2019-05-02', '2019-05-03'])).must_equal true
+    expect(DateRange.dates_overlap?(['2019-05-01', '2019-05-02', '2019-05-03'], ['2019-05-02', '2019-05-03', '2019-05-04'])).must_equal true
+    expect(DateRange.dates_overlap?(['2019-05-01', '2019-05-02', '2019-05-03'], ['2019-05-03', '2019-05-04', '2019-05-05'])).must_equal false
```
Comment (droberts-sea) [inline spec/date_range_spec.rb:16]: These are very hard to read! You might break each of these out into a separate test with a descriptive name. This has the added benefit of, if one of them fails, you know exactly which one it was.

File: lib/date_range.rb [hunk 11]
```diff
@@ -0,0 +1,21 @@
+require 'time'
+
+class DateRange
+  # Check if given date is valid 
+  def self.valid_date?(date)
+    begin
```
Comment (droberts-sea) [inline lib/date_range.rb:6]: I like that you've isolated the logic for working with date ranges in this class - good organization.

I am curious about your decision to make these all self methods, rather than creating instances of a `DateRange` class. For an alternative way to approach it, you might check out the [instructor implementation](https://github.com/droberts-ada/hotel/blob/dpr/solution/lib/date_range.rb).

**Comment by droberts-sea** [comment]
# Hotel
## What We're Looking For



Feature | Feedback
---     | ---
**Baseline** |
Used git regularly | yes
Answer comprehension questions | yes
**Design** |
Each class is responsible for a single piece of the program | yes
Classes are loosely coupled | yes
**Wave 1** |
List rooms | yes
Reserve a room for a given date range | yes
List reservations for a given date | yes
Calculate reservation price | yes
Invalid date range produces an error | no - see inline
Test coverage | yes
**Wave 2** |
View available rooms for a given date range | yes
Reserving a room that is not available produces an error | yes
Test coverage | yes
**Wave 3** |
Create a block of rooms | good start
Check if a block has rooms | good start
Reserve a room from a block | good start
Test coverage | good start
**Fundamentals** |
Names variables, classes and modules appropriately | yes
Understanding of variable scope - local vs instance | yes
Can create complex logical structures utilizing variables | yes
Appropriately uses methods to break down tasks into smaller simpler tasks | yes
Understands the differences between class and instance methods | yes
Appropriately uses iterators and `Enumerable` methods | yes
Appropriately writes and utilizes classes | yes
Appropriately utilizes modules as a namespace | no 
**Wrap Up** |
There is a refactors.txt file | yes
The file provides a roadmap to future changes | yes
**Additional Feedback** | Good job overall! This is a big difficult project, and you've done good work figuring out a design, organizing your ideas into methods that make sense, and building an implementation that works. While I've left a number of nitpicks inline, it is clear to me that the learning goals for this assignment were met. Keep up the hard work!
